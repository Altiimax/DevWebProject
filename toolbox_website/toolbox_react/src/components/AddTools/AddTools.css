.baseForm {
  background-color: white;
  padding: 10px;
  border-radius: 0px 0px 10px 10px;
}

.addToolForm {
  color: black;
  background-color: rgb(255, 255, 255);
  border-radius: 10px;
  padding: 20px;
  margin: 1em;
}

.formTitle {
  border-radius: 10px 10px 0px 0px;
  color: black;
  margin-bottom: 0;
  text-align: center;
  padding: 10px;
}

.FormField_TextArea {
  font-family: "Abel", sans-serif;
  padding: 0;
  margin: 0em 0em 1.5em 0em;
  color: black;
  width: 85%;
  background-color: transparent;
  border: none;
  outline: none;
  border: 1px solid #445366;
  font-size: 1em;
  font-weight: 300;
}
