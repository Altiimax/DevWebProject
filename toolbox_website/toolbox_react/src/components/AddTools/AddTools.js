import React, { Component } from "react";
import PropTypes from "prop-types";
import { Modal } from "react-bootstrap";
import { apiRequest } from "../../api/apiRequest.js";
import { userFromToken } from "../../utils";
import addimg from "./addimg.png";

import "./AddTools.css";

const initialStates = {
  toolName: "",
  toolDescription: "",
  price: 0.00,
  typeOfRent: "",
  pictures: [],
};

let pictures_preview = [addimg,addimg,addimg,addimg,addimg];

/**
 *This component is used to register the information about a new tool added by
 * the user on the database. He must enter the tool name,
 * select a type of renting (by default it's per day), a currency and the price.
 * He can also add a description of the tool and pictures of it.
 * @param
 */
class AddTools extends Component {
  constructor(props) {
    super(props);
    this.state = initialStates;
    this.showPopUp = true;
  }

  componentDidUpdate() {
    this.showPopUp = this.props.showPopUp;
  }

  closePopUp = () => {
    this.showPopUp = false;
    this.setState(initialStates);
    pictures_preview = [addimg,addimg,addimg,addimg,addimg];
  };

  addToolApi = (user_id, data) => {
    let endpoint = "/api/persons/" + user_id + "/tools/";
    let req = new apiRequest();
    req.open("POST", `${endpoint}`);
    req.contentType("json");

    let self = this;
    req.addEventListener("readystatechange", function () {
      if (this.readyState === 4) {
        if (this.status === 201) {
          let toolId = JSON.parse(this.responseText).id_tool;
          let dataImages = new FormData();
          for( let img of self.state.pictures){
            dataImages.append("image", img, img.name);
            self.addToolImagesApi(toolId,dataImages);
          }
        }
      }
    });
    req.send(data);
  };

  addToolImagesApi = (tool_id, data) => {
    let endpoint = "/api/tools/" + tool_id + "/images/";
    let req = new apiRequest();
    req.open("POST", `${endpoint}`);
    req.contentType("formData");

    let self = this;

    req.addEventListener("readystatechange", function () {
      if (this.readyState === 4) {
        if (this.status === 201) {
          self.props.refreshToolList();
        }
      }
    });
    req.send(data);
  };

  handleChange = (e) => {
    let target = e.target;
    let value = target.type === "file" ? target.files[0] : target.value;
    let name = target.name;
    if(target.type === "file"){
      let imgPos = parseInt(name.slice(-1))-1;
      pictures_preview[imgPos] = URL.createObjectURL(value);
      const prevPictures = this.state.pictures;
      prevPictures[imgPos] = value;
      this.setState({ pictures: prevPictures });
    }
    else{
      this.setState({
        [name]: value,
      });
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    let dataTool = {
      toolName: this.state.toolName,
      toolDescription: this.state.toolDescription,
      toolPrice: this.state.price,
    };
    if(this.state.pictures.length === 0){
      document.getElementById("imgError").style.display="block";
    }
    else{
      this.addToolApi(userFromToken().id, JSON.stringify(dataTool));
      this.closePopUp();
    }
  };

  render() {
    return (
      <>
      <Modal show={this.showPopUp}
          onHide={() => {
            this.closePopUp();
          }}
      >
      <Modal.Header closeButton>
        <Modal.Title>Add new tool</Modal.Title>
      </Modal.Header>

      <Modal.Body>
        <form className="addToolForm" onSubmit={this.handleSubmit}>
          <label className="FormField_Label" htmlFor="toolName">
            Tool Name
          </label>
          <input
            required
            type="text"
            className="FormField_Input"
            name="toolName"
            placeholder="Enter the tool name"
            onChange={this.handleChange}
          />

          <label className="FormField_Label" htmlFor="toolDescription">
            Tool Description
          </label>
          <textarea
            type="text"
            className="FormField_TextArea"
            name="toolDescription"
            onChange={this.handleChange}
          />
          
          <div className="FormRow">
            <div>
              <label className="FormField_Label" htmlFor="price">
                Price
              </label>
              <div id="priceInput">
                <input
                  className="FormField_Input"
                  required
                  type="number"
                  min="0"
                  max="99999.99"
                  step="0.01"
                  name="price"
                  placeholder="0.00"
                  onChange={this.handleChange}
                />
                <div>â‚¬/day</div>
              </div>
            </div>

            {/*<div>  
              <label className="FormField_Label" htmlFor="typeOfRent">
                Type of renting
              </label>
              <select defaultValue="Per day" name="typeOfRent">
                <option value="Per hours">Per hours</option>
                <option value="Per day">Per day</option>
                <option value="Per week">Per week</option>
                <option value="Per month">Per month</option>
              </select>
            </div>*/}
          </div>
          
          <label className="FormField_Label">Tool Images</label>
          <div className="FormImgsDrop">
            <label htmlFor="picture_1">
              <img className="addImgBtn" src={pictures_preview[0]} alt="add_image"/>
              <input
                type="file"
                id="picture_1"
                name="picture_1"
                style={{visibility:"hidden"}}
                onChange={this.handleChange}
              />
            </label>
            <label htmlFor="picture_2">
              <img className="addImgBtn" src={pictures_preview[1]} alt="add_image"/>
              <input
                type="file"
                id="picture_2"
                name="picture_2"
                style={{visibility:"hidden"}}
                onChange={this.handleChange}
              />
            </label>
            <label htmlFor="picture_3">
              <img className="addImgBtn" src={pictures_preview[2]} alt="add_image"/>
              <input
                type="file"
                id="picture_3"
                name="picture_3"
                style={{visibility:"hidden"}}
                onChange={this.handleChange}
              />
            </label>
            <label htmlFor="picture_4">
              <img className="addImgBtn" src={pictures_preview[3]} alt="add_image"/>
              <input
                type="file"
                id="picture_4"
                name="picture_4"
                style={{visibility:"hidden"}}
                onChange={this.handleChange}
              />
            </label>
            <label htmlFor="picture_5">
              <img className="addImgBtn" src={pictures_preview[4]} alt="add_image"/>
              <input
                type="file"
                id="picture_5"
                name="picture_5"
                style={{visibility:"hidden"}}
                onChange={this.handleChange}
              />
            </label>            
          </div>

          <div id="imgError" className="error">You must upload at least one image!</div>

          <div className="FormBtns">
            <input className="FormCancelBtn" type="button" value="Cancel" onClick={this.closePopUp}/>
            <input className="FormSubmitBtn" type="submit" value="Add tool" />
          </div>
        </form>
      </Modal.Body>

      </Modal>
      </>
    );
  }
}

export default AddTools;

AddTools.propTypes = {
  showPopUp: PropTypes.bool.isRequired,
  refreshToolList: PropTypes.func.isRequired,
};