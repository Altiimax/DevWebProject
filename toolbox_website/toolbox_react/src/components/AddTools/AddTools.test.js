import React from "react";
import ReactDOM from "react-dom";
import AddTools from "./AddTools.js";
import TestRenderer from "react-test-renderer";
test("render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<AddTools />, div);
});

/*
Test de la snapshot (a faire que si le component ne bouge pas trop dans le temps)
Ce test ajoute un dossier _snaphsot_dans lequel est contenu toutes les snapshots
*/
it("renders a form", () => {
  const testRenderer = TestRenderer.create(<AddTools />);
  const testInstance = testRenderer.toJSON();
  expect(testInstance).toMatchSnapshot();
});
