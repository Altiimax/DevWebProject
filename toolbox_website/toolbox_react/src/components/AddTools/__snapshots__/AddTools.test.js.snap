// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a form 1`] = `
<div
  className="addToolForm"
>
  <h1
    className="formTitle"
  >
    Add new tool
  </h1>
  <form
    className="baseForm"
  >
    <label
      className="FormField_Label"
      htmlFor="toolName"
    >
      Tool Name
    </label>
    <input
      className="FormField_Input"
      name="toolName"
      onChange={[Function]}
      placeholder="Enter the tool name"
      required={true}
      type="text"
    />
    <label
      className="FormField_Label"
      htmlFor="toolDescription"
    >
      Tool Description
    </label>
    <textarea
      className="FormField_TextArea"
      name="toolDescription"
      onChange={[Function]}
      type="text"
    />
    <label
      className="FormField_Label"
      htmlFor="toolState"
    >
      State of wear
    </label>
    <select
      defaultValue="Used"
      name="toolState"
    >
      <option
        value="New"
      >
        New
      </option>
      <option
        value="Bit used"
      >
        Bit used
      </option>
      <option
        value="Used"
      >
        Used
      </option>
      <option
        value="Pretty used"
      >
        Pretty used
      </option>
      <option
        value="In bad conditions"
      >
        In bad conditions
      </option>
    </select>
    <div
      className="form-row"
    >
      <div
        className="col"
      >
        <label
          className="FormField_Label"
          htmlFor="price"
        >
          Price
        </label>
        <input
          max="99999.99"
          min="0"
          name="price"
          onChange={[Function]}
          required={true}
          step="0.01"
          type="number"
        />
      </div>
      <div
        className="col"
      >
        <label
          className="FormField_Label"
          htmlFor="currency"
        >
          Currency
        </label>
        <input
          name="currency"
          onChange={[Function]}
          required={true}
          type="text"
        />
      </div>
      <div
        className="col"
      >
        <label
          className="FormField_Label"
          htmlFor="typeOfRent"
        >
          Type of renting
        </label>
        <select
          defaultValue="Per day"
          name="typeOfRent"
        >
          <option
            value="Free"
          >
            Free
          </option>
          <option
            value="Per hours"
          >
            Per hours
          </option>
          <option
            value="Per day"
          >
            Per day
          </option>
          <option
            value="Per week"
          >
            Per week
          </option>
          <option
            value="Per month"
          >
            Per month
          </option>
        </select>
      </div>
    </div>
    <br />
    <div
      className="form-row"
    >
      <label
        htmlFor="pictures"
      >
        Select a file: 
      </label>
      <input
        id="pictures"
        name="pictures"
        onChange={[Function]}
        type="file"
      />
    </div>
    <br />
    <div
      className="FormBtns"
    >
      <input
        className="FormCancelBtn"
        type="button"
        value="Cancel"
      />
      <input
        className="FormSubmitBtn"
        type="submit"
        value="Add tool"
      />
    </div>
  </form>
</div>
`;
